File: src/qrs-core-python/src/qrs/id.py
"""Identifier and encoding helpers.

All protocol identifiers are 128-bit truncated SHA-256 digests. This module is
deliberately platform-agnostic and delegates every cryptographic operation to
Python's standard library (:mod:`hashlib`, :mod:`secrets`) rather than
hand-rolled code.
"""

from __future__ import annotations

import base64
import hashlib
import re
import secrets
from typing import Final

from .errors import QrsError, QrsParseError

# Number of identifier bytes used for all truncated identifiers.
ID_BYTES: Final = 16

_HEX_RE = re.compile(r"^[0-9a-fA-F]*$")

__all__ = [
    "ID_BYTES",
    "sha256",
    "sha384",
    "sha3_512",
    "hash_for",
    "is_hash_algorithm",
    "trunc_sha256",
    "constant_time_equal",
    "to_hex",
    "from_hex",
    "to_base64url",
    "from_base64url",
    "random_bytes",
    "random_id",
    "uint8_to_int",
    "int_to_uint8",
    "concat_bytes",
]


def sha256(data: bytes) -> bytes:
    """Full SHA-256 digest of arbitrary bytes."""
    return hashlib.sha256(data).digest()


def sha384(data: bytes) -> bytes:
    """Full SHA-384 digest of arbitrary bytes."""
    return hashlib.sha384(data).digest()


def sha3_512(data: bytes) -> bytes:
    """Full SHA3-512 digest of arbitrary bytes."""
    return hashlib.sha3_512(data).digest()


def is_hash_algorithm(value: object) -> bool:
    """True when *value* is one of the supported hash-algorithm names."""
    return value in ("SHA-256", "SHA-384", "SHA3-512")


def hash_for(alg: str | None, data: bytes) -> bytes:
    """Hash arbitrary bytes with the given algorithm (defaults to SHA-256)."""
    if alg == "SHA-384":
        return sha384(data)
    if alg == "SHA3-512":
        return sha3_512(data)
    return sha256(data)


def trunc_sha256(data: bytes, length: int = ID_BYTES) -> bytes:
    """Truncated SHA-256 digest (defaults to the 128-bit protocol identifier)."""
    return sha256(data)[:length]


def constant_time_equal(a: str, b: str) -> bool:
    """Constant-time string equality (used for bound-value comparisons)."""
    if len(a) != len(b):
        return False
    import hmac

    return hmac.compare_digest(a.encode("utf-8"), b.encode("utf-8"))


def to_hex(data: bytes) -> str:
    """Hex-encode bytes (lowercase)."""
    return data.hex()


def from_hex(hex_string: str) -> bytes:
    """Hex-decode a string; raises :class:`QrsParseError` on malformed input."""
    if not isinstance(hex_string, str) or not _HEX_RE.match(hex_string) or len(hex_string) % 2 != 0:
        raise QrsParseError("Invalid hex string")
    return bytes.fromhex(hex_string)


def to_base64url(data: bytes) -> str:
    """Base64url (no padding) encode bytes."""
    return base64.urlsafe_b64encode(data).rstrip(b"=").decode("ascii")


def from_base64url(value: str) -> bytes:
    """Base64url decode; tolerates standard base64 characters and padding."""
    normalized = value.replace("+", "-").replace("/", "_").rstrip("=")
    try:
        # urlsafe_b64decode accepts unpadded input.
        return base64.urlsafe_b64decode(normalized + "=" * (-len(normalized) % 4))
    except Exception as exc:  # noqa: BLE001 - normalize any decoding failure
        raise QrsParseError("Invalid base64url string") from exc


def random_bytes(length: int) -> bytes:
    """Generate *length* cryptographically secure random bytes."""
    return secrets.token_bytes(length)


def random_id(num_bytes: int = ID_BYTES) -> str:
    """Generate a random hex identifier (defaults to 16 bytes)."""
    return to_hex(random_bytes(num_bytes))


def uint8_to_int(data: bytes) -> int:
    """Interpret bytes as a big-endian unsigned integer."""
    return int.from_bytes(data, "big")


def int_to_uint8(value: int, length: int = 8) -> bytes:
    """Encode an integer as big-endian bytes of *length* bytes."""
    return value.to_bytes(length, "big")


def concat_bytes(*parts: bytes) -> bytes:
    """Concatenate byte strings."""
    return b"".join(parts)


def require_secure_random() -> None:
    """Raise if the platform cannot provide secure randomness (never happens on CPython)."""
    try:
        secrets.token_bytes(1)
    except NotImplementedError as exc:  # pragma: no cover - exotic platforms only
        raise QrsError("No secure random source available") from exc

File: qrs-core-js/src/id.ts
/**
 * Identifier and encoding helpers.
 *
 * All protocol identifiers are 128-bit truncated SHA-256 digests. This module is
 * deliberately platform-agnostic (no Node imports, no Buffer) and delegates every
 * cryptographic operation to audited, well-maintained libraries rather than
 * hand-rolled code:
 *   - SHA-256: `@noble/hashes` (audited, dependency-free, works in Node, browsers
 *     and React Native);
 *   - hex / base64url: `@noble/hashes/utils` and `@scure/base`;
 *   - random bytes: `globalThis.crypto.getRandomValues` (available in Node >= 20,
 *     browsers, and React Native when a polyfill such as
 *     `react-native-get-random-values` is installed).
 */
import { sha256 as nobleSha256, sha384 as nobleSha384 } from '@noble/hashes/sha2.js';
import { sha3_512 as nobleSha3_512 } from '@noble/hashes/sha3.js';
import { bytesToHex, hexToBytes } from '@noble/hashes/utils.js';
import { base64urlnopad } from '@scure/base';
import { QrsError, QrsParseError } from './errors.js';

/** Number of identifier bytes used for all truncated identifiers. */
export const ID_BYTES = 16;

/** Full SHA-256 digest of arbitrary bytes (synchronous, portable). */
export function sha256(data: Uint8Array): Uint8Array {
  return nobleSha256(data);
}

/** Hash algorithms supported for TCert-bound content hashing (attachments). */
export type HashAlgorithm = 'SHA-256' | 'SHA-384' | 'SHA3-512';

export const HASH_ALGORITHMS: readonly HashAlgorithm[] = ['SHA-256', 'SHA-384', 'SHA3-512'];

export function isHashAlgorithm(value: unknown): value is HashAlgorithm {
  return HASH_ALGORITHMS.includes(value as HashAlgorithm);
}

/** Full SHA-384 digest of arbitrary bytes. */
export function sha384(data: Uint8Array): Uint8Array {
  return nobleSha384(data);
}

/** Full SHA3-512 digest of arbitrary bytes. */
export function sha3_512(data: Uint8Array): Uint8Array {
  return nobleSha3_512(data);
}

/** Hash arbitrary bytes with the given algorithm (defaults to SHA-256). */
export function hashFor(alg: HashAlgorithm | undefined, data: Uint8Array): Uint8Array {
  switch (alg ?? 'SHA-256') {
    case 'SHA-384':
      return sha384(data);
    case 'SHA3-512':
      return sha3_512(data);
    default:
      return sha256(data);
  }
}

/** Truncated SHA-256 digest (defaults to the 128-bit protocol identifier). */
export function truncSha256(data: Uint8Array, bytes = ID_BYTES): Uint8Array {
  const full = sha256(data);
  return full.slice(0, Math.min(bytes, full.length));
}

/** Constant-time string equality (used for bound-value comparisons). */
export function constantTimeEqual(a: string, b: string): boolean {
  if (a.length !== b.length) return false;
  let diff = 0;
  for (let i = 0; i < a.length; i++) diff |= a.charCodeAt(i) ^ b.charCodeAt(i);
  return diff === 0;
}

/* -------------------------------------------------------------------------- */
/* Encoding helpers (no Buffer dependency)                                    */
/* -------------------------------------------------------------------------- */

export function toHex(bytes: Uint8Array): string {
  return bytesToHex(bytes);
}

export function fromHex(hex: string): Uint8Array {
  if (!/^[0-9a-fA-F]*$/.test(hex) || hex.length % 2 !== 0) {
    throw new QrsParseError('Invalid hex string');
  }
  return hexToBytes(hex);
}

export function toBase64Url(bytes: Uint8Array): string {
  return base64urlnopad.encode(bytes);
}

export function fromBase64Url(value: string): Uint8Array {
  // Tolerate standard base64 characters and optional padding, then delegate.
  const normalized = value.replace(/\+/g, '-').replace(/\//g, '_').replace(/=+$/, '');
  try {
    return base64urlnopad.decode(normalized);
  } catch {
    throw new QrsParseError('Invalid base64url string');
  }
}

/* -------------------------------------------------------------------------- */
/* Secure random                                                              */
/* -------------------------------------------------------------------------- */

/** Generate `length` cryptographically secure random bytes (portable). */
export function randomBytes(length: number): Uint8Array {
  const cryptoObj = (globalThis as { crypto?: { getRandomValues?: (a: Uint8Array) => Uint8Array } }).crypto;
  if (!cryptoObj || typeof cryptoObj.getRandomValues !== 'function') {
    throw new QrsError('No secure random source available (globalThis.crypto.getRandomValues missing)');
  }
  const out = new Uint8Array(length);
  cryptoObj.getRandomValues(out);
  return out;
}

/** Generate a random hex identifier (defaults to 16 bytes). */
export function randomId(bytes = ID_BYTES): string {
  return toHex(randomBytes(bytes));
}

/** Hex encode a big-endian uint8. */
export function uint8ToBigInt(bytes: Uint8Array): bigint {
  let out = 0n;
  for (const b of bytes) out = (out << 8n) | BigInt(b);
  return out;
}

